design/build-cache.md

```diff
@@ -0,0 +1,201 @@
+- Have a single commit that gets deployed from a single PR or branch with multiple configurations.
+
+Scoping should therefore take into account:
+
+1. Branch
+2. Commit
+3. Hash signature of content to be cached
+
+#### Job level
+
+This level of caching explicitly links a cache to certain jobs.
+
+Use when:
+
+1. Injecting the cached resources into every job is not desirable.
+
+#### Event level
+
+Event level scoping is scoped to a single pipeline execution event. Cached data is visible and accessible to all jobs within a single execution of the pipeline.
+
+Executions of a pipeline for the same commit are viewed as different executions.
+
+Use when:
+
+1. Generated resources are time dependant
+2. Generated resources are only valid for single execution
+3. Generated resources are meant for deployment (Please note that triggering deployment with a detached job will create a new event and, by extension, cache.)
+
+#### Pipeline level
+
+Pipeline level scoping is scoped to a pipeline as a whole. Cached data is visible and accessible to all instances of the pipeline.
+
+Note: Concurrency for simultaneous builds need proper attention.
+
+Use when:
+
+1. Build dependencies take a long time to resolve
+2. Generated resources are meant for deployment (Risk prone, not recommended unless deleting deployed artifact from cache after deployment)
+
+### (Proposal) Local and remote caching
+
+Pipeline and event level caches are stored on the build cache service node. They are copied to the build node on pipeline setup, and back to the build cache on pipeline completion.
+
+Once copied to the build node, they are persisted as a shared volume, mounted onto the appropriate path in the build container. 
+
+This proposal is currently blocked by the nature of the backend implementation for nodes. The k8s executor for example, selects a completely random node for every job in the pipeline. Caching locally is therefore not viable at the moment.
+
+### Build cache service
+
+The build cache service is essentially a specialized version of the store service. It utilizes the builtin Hapi cache system backed by Catbox. 
+
+### Implementation details
+
+#### Execution flow
+
+![flow.png](./diagrams/flow.png)
+
+#### Server
+
+##### API
+
+| Method | URL | Body | Reponse |
+| - | - | - | - |
+| GET | /pipeline/:pipeline_id/cache/:cache_id | none | A single cache entity.<br><b>Body</b><br>```{ "scope": "", "event_id": "", "pipeline_id": "", "ttl": 100,"data": base64}``` <br><b>Codes</b><br> 200, 404 |
+| GET | /pipeline/:pipeline_id/cache | none | List of currently valid caches for the pipeline.<br><b>Body</b><br>```[{ "scope": "", "event_id": "", "pipeline_id": "", "ttl": 100,"data": base64}]``` <br><b>Codes</b><br> 200, 404 | 
+| PUT | /pipeline/:pipeline_id/cache | ```{ "scope": "", "event_id": "", "pipeline_id": "", "data": base64``` | <b>Codes</b><br>202, 403 |
+
+Points to consider:
+
+- Should maybe just use GraphQL?
```

Comment:
I doubt Hapi supports this. Can I scrap this?

Reply:
We don't have to use Hapi to implement the cache server if it doesn't fit the design (unless we plan to reuse the store API). But I'm actually not very familiar with GraphQL... What's the reason of using GraphQL instead of REST?